### PR TITLE
Share mojom types between C++ and wallet desktop frontend

### PR DESCRIPTION
**Pull request by yrliou**
Resolves https://github.com/brave/brave-browser/issues/19247

PR summary:
- Refactor wallet_api_proxy, wallet_page_api_proxy, wallet_panel_api_proxy to typescript, and update callers for getting these proxies.
- Remove duplicate type definitions in constants/types.ts and use mojom types directly.

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by petemill — APPROVED** [review]
Great job. All my comments are minor and could be done as follow-ups.

**Comment by petemill** [comment]
There's a missing dep in `.storybook/BUILD.gn` - it needs to make sure the mojo generated JS is built, please see existing VPN example in that file.

**Comment by yrliou** [comment]
> There's a missing dep in `.storybook/BUILD.gn` - it needs to make sure the mojo generated JS is built, please see existing VPN example in that file. again: ok for follow-up as it won't break CI at the moment since brave_wallet build flag is enabled on all platforms I think

Thanks, this is addressed in https://github.com/brave/brave-core/pull/10967/commits/8a477cbbda609a9f3f142e19baf3d41e53cc6730.

File: components/brave_wallet_ui/panel/BUILD.gn [hunk 1]
```diff
@@ -34,13 +33,5 @@ preprocess_if_expr("preprocess_panel") {
...
-  in_files = [ "wallet_panel_api_proxy.js" ]
-}
-
-js_library("wallet_panel_api_proxy") {
-  deps = [
-    "//brave/components/brave_wallet/common:mojom_js_library_for_compile",
-    "//ui/webui/resources/js:cr.m",
-    "//url/mojom:url_mojom_gurl_js_library_for_compile",
-  ]
+  in_files = [ "wallet_panel_api_proxy.ts" ]
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet_ui/panel/BUILD.gn]: Addressed in https://github.com/brave/brave-core/pull/10967/commits/8a477cbbda609a9f3f142e19baf3d41e53cc6730

File: components/brave_wallet_ui/page/async/wallet_page_async_handler.ts [hunk 2]
```diff
@@ -166,11 +166,11 @@ handler.on(WalletPageActions.importFromCryptoWallets.getType(), async (store: St
   const keyringController = (await getAPIProxy()).keyringController
   const result = await braveWalletService.importFromCryptoWallets(payload.password, payload.newPassword)
   if (result.success) {
-    await keyringController.notifyWalletBackupComplete()
+    keyringController.notifyWalletBackupComplete()
   }
   store.dispatch(WalletPageActions.setImportWalletError({
     hasError: !result.success,
-    errorMessage: result.errorMessage
+    errorMessage: result.errorMessage ? result.errorMessage : undefined
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet_ui/page/async/wallet_page_async_handler.ts]: Addressed in https://github.com/brave/brave-core/pull/10967/commits/8a477cbbda609a9f3f142e19baf3d41e53cc6730

File: components/brave_wallet_ui/page/BUILD.gn [hunk 3]
```diff
@@ -35,13 +34,5 @@ preprocess_if_expr("preprocess_page") {
...
-  in_files = [ "wallet_page_api_proxy.js" ]
-}
-
-js_library("wallet_page_api_proxy") {
-  deps = [
-    "//brave/components/brave_wallet/common:mojom_js_library_for_compile",
-    "//ui/webui/resources/js:cr.m",
-    "//url/mojom:url_mojom_gurl_js_library_for_compile",
-  ]
+  in_files = [ "wallet_page_api_proxy.ts" ]
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet_ui/page/BUILD.gn]: Addressed in https://github.com/brave/brave-core/pull/10967/commits/8a477cbbda609a9f3f142e19baf3d41e53cc6730

File: components/brave_wallet_ui/common/async/lib.ts [hunk 4]
```diff
@@ -47,11 +50,13 @@ export const onConnectHardwareWallet = (opts: HardwareWalletConnectOpts): Promis
...
     const apiProxy = await getAPIProxy()
     const keyring = await apiProxy.getKeyringsByType(opts.hardware)
-    keyring.getAccounts(opts.startIndex, opts.stopIndex, opts.scheme)
-      .then(async (accounts: HardwareWalletAccount[]) => {
-        resolve(accounts)
-      })
-      .catch(reject)
+    if (keyring instanceof LedgerBridgeKeyring || keyring instanceof TrezorBridgeKeyring) {
+      keyring.getAccounts(opts.startIndex, opts.stopIndex, opts.scheme)
+        .then(async (accounts: HardwareWalletAccount[]) => {
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet_ui/common/async/lib.ts]: Addressed in https://github.com/brave/brave-core/pull/10967/commits/8a477cbbda609a9f3f142e19baf3d41e53cc6730

File: components/brave_wallet_ui/common/async/handlers.ts [hunk 5]
```diff
@@ -313,12 +320,9 @@ handler.on(WalletActions.sendTransaction.getType(), async (store: Store, payload
       payload.value,
       payload.data || []
     )
+    addResult = await (apiProxy.ethTxController.addUnapprovedTransaction(txData, payload.from))
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet_ui/common/async/handlers.ts]: Addressed in https://github.com/brave/brave-core/pull/10967/commits/8a477cbbda609a9f3f142e19baf3d41e53cc6730

File: components/brave_wallet_ui/common/async/handlers.ts [hunk 6]
```diff
@@ -301,7 +306,9 @@ handler.on(WalletActions.sendTransaction.getType(), async (store: Store, payload
       payload.value,
       payload.data || []
     )
-    : apiProxy.makeTxData(
+    addResult = await (apiProxy.ethTxController.addUnapproved1559Transaction(txData, payload.from))
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet_ui/common/async/handlers.ts]: Addressed in https://github.com/brave/brave-core/pull/10967/commits/8a477cbbda609a9f3f142e19baf3d41e53cc6730

File: components/brave_wallet_ui/utils/datetime-utils.ts [hunk 7]
```diff
@@ -1,11 +1,16 @@
...
+
+import { TimeDelta } from 'gen/mojo/public/mojom/base/time.mojom.m.js'
 
 /**
  * Converts a mojo time to a JS time.
  * @param {!mojoBase.mojom.TimeDelta} mojoTime
  * @return {!Date}
  */
-export function convertMojoTimeToJS (mojoTime: MojoTime) {
+export function convertMojoTimeToJS (mojoTime: TimeDelta) {
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet_ui/utils/datetime-utils.ts]: Addressed in https://github.com/brave/brave-core/pull/10967/commits/8a477cbbda609a9f3f142e19baf3d41e53cc6730
Haven't moved into mojomUtils.ts to keep this PR simpler, will move it in a follow-up.

File: components/brave_wallet_ui/common/async/handlers.ts [hunk 8]
```diff
@@ -52,6 +52,9 @@ import {
 } from './lib'
 import { Store } from './types'
 
+import LedgerBridgeKeyring from '../../common/ledgerjs/eth_ledger_bridge_keyring'
+import TrezorBridgeKeyring from '../../common/trezor/trezor_bridge_keyring'
+
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet_ui/common/async/handlers.ts:57]: Looks like we're not following this rule in this file already before my change, will open a follow-up issue for this.

**Review by Douglashdaniel — APPROVED** [review]
++

File: components/brave_wallet_ui/utils/datetime-utils.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet_ui/utils/datetime-utils.ts]: created https://github.com/brave/brave-browser/issues/19345

File: components/brave_wallet_ui/common/async/bridge.ts [hunk 9]
```diff
@@ -3,14 +3,14 @@
...
   let api
   if (window.location.hostname === 'wallet-panel.top-chrome') {
-    api = await import('../../panel/wallet_panel_api_proxy.js')
+    api = await import('../../panel/wallet_panel_api_proxy')
   } else {
-    api = await import('../../page/wallet_page_api_proxy.js')
+    api = await import('../../page/wallet_page_api_proxy')
   }
-  return api.default.getInstance()
+  return api.default()
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet_ui/common/async/bridge.ts:15]: https://github.com/brave/brave-browser/issues/19346

File: components/brave_wallet_ui/common/wallet_api_proxy.ts [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+  }
+
+  makeEIP1559TxData (chainId: string, nonce: string, maxPriorityFeePerGas: string, maxFeePerGas: string, gasLimit: string, to: string, value: string, data: number[]) {
+    const txData = new BraveWallet.TxData1559()
+    txData.baseData = this.makeTxData(nonce, '', gasLimit, to, value, data)
+    txData.maxPriorityFeePerGas = maxPriorityFeePerGas
+    txData.maxFeePerGas = maxFeePerGas
+    txData.chainId = chainId
+    return txData
+  }
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet_ui/common/wallet_api_proxy.ts:60]: https://github.com/brave/brave-browser/issues/19347

File: components/brave_wallet_ui/panel/wallet_panel_api_proxy.ts [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    this.panelHandler.closeUI()
+  }
+
+  connectToSite (accounts: string[], origin: string, tabId: number) {
+    this.panelHandler.connectToSite(accounts, origin, tabId)
+  }
+
+  cancelConnectToSite (origin: string, tabId: number) {
+    this.panelHandler.cancelConnectToSite(origin, tabId)
+  }
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet_ui/panel/wallet_panel_api_proxy.ts:43]: https://github.com/brave/brave-browser/issues/19347

**Pull request merged**